test/types/route.test-d.ts

```diff
@@ -0,0 +1,133 @@
+import {expectType} from 'tsd'
+
+/*
+ * Testing Fastify HTTP Routes and Route Shorthands.
+ * Verifies Request and Reply types as well.
+ * For the route shorthand tests the argument orders are:
+ * - `(path, handler)`
+ * - `(path, options, handler)`
+ * - `(path, options)`
+ */
+
+const routeHandler: RouteHandlerMethod = (request, reply) => {
+  expectType<FastifyRequest>(request)
+  expectType<FastifyReply>(reply)
+}
+
+// route method
+expectType<FastifyInstance>(fastify().route({
+  method: 'GET',
+  url: '/',
+  handler: routeHandler
+}))
+expectType<FastifyInstance>(fastify().route({
+  method: 'POST',
+  url: '/',
+  handler: routeHandler
+}))
+expectType<FastifyInstance>(fastify().route({
+  method: 'PUT',
+  url: '/',
+  handler: routeHandler
+}))
+expectType<FastifyInstance>(fastify().route({
+  method: 'PATCH',
+  url: '/',
+  handler: routeHandler
+}))
+expectType<FastifyInstance>(fastify().route({
+  method: 'HEAD',
+  url: '/',
+  handler: routeHandler
+}))
+expectType<FastifyInstance>(fastify().route({
+  method: 'DELETE',
+  url: '/',
+  handler: routeHandler
+}))
+expectType<FastifyInstance>(fastify().route({
+  method: 'OPTIONS',
+  url: '/',
+  handler: routeHandler
+}))
+
+// get
+expectType<FastifyInstance>(fastify().get('/', routeHandler))
+expectType<FastifyInstance>(fastify().get('/', {}, routeHandler))
+expectType<FastifyInstance>(fastify().get('/', { handler: routeHandler }))
+
+// post
+expectType<FastifyInstance>(fastify().post('/', routeHandler))
+expectType<FastifyInstance>(fastify().post('/', {}, routeHandler))
+expectType<FastifyInstance>(fastify().post('/', { handler: routeHandler }))
+
+// put
+expectType<FastifyInstance>(fastify().put('/', routeHandler))
+expectType<FastifyInstance>(fastify().put('/', {}, routeHandler))
+expectType<FastifyInstance>(fastify().put('/', { handler: routeHandler }))
+
+// patch
+expectType<FastifyInstance>(fastify().patch('/', routeHandler))
+expectType<FastifyInstance>(fastify().patch('/', {}, routeHandler))
+expectType<FastifyInstance>(fastify().patch('/', { handler: routeHandler }))
+
+// head
+expectType<FastifyInstance>(fastify().head('/', routeHandler))
+expectType<FastifyInstance>(fastify().head('/', {}, routeHandler))
+expectType<FastifyInstance>(fastify().head('/', { handler: routeHandler }))
+
+// delete
+expectType<FastifyInstance>(fastify().delete('/', routeHandler))
+expectType<FastifyInstance>(fastify().delete('/', {}, routeHandler))
+expectType<FastifyInstance>(fastify().delete('/', { handler: routeHandler }))
+
+// options
+expectType<FastifyInstance>(fastify().options('/', routeHandler))
+expectType<FastifyInstance>(fastify().options('/', {}, routeHandler))
+expectType<FastifyInstance>(fastify().options('/', { handler: routeHandler }))
+
+// all
+expectType<FastifyInstance>(fastify().all('/', routeHandler))
+expectType<FastifyInstance>(fastify().all('/', {}, routeHandler))
+expectType<FastifyInstance>(fastify().all('/', { handler: routeHandler }))
+
```

Comment:
I suggest adding a test ensuring with an unsupported method such as 'CONNECT' or a typo 'OPTION' ensuring that this raise an error.

Reply:
You should add a test with an array of method such as ['GET', 'POST'].
Ref : https://www.fastify.io/docs/latest/Routes/
> method: currently it supports 'DELETE', 'GET', 'HEAD', 'PATCH', 'POST', 'PUT' and 'OPTIONS'. It could also be an array of methods.

Reply:
> I suggest adding a test ensuring with an unsupported method such as 'CONNECT' or a typo 'OPTION' ensuring that this raise an error.

I should be able to test that within context of the `.route` method. But otherwise it will be difficult as the `.decorate` utility allows developers to add whatever method they want to the fastify instance.